src/connection.rs

```diff
@@ -0,0 +1,188 @@
+                    // If there is a new channel id we update it
+                    if data.has_channel_id() {
+                        self.users[index].set_channel_id(data.get_channel_id());
+                    }
+                } else {
+                    // Add user
+                    self.users.push(*data.clone());
+                }
+            }
+            // Ignore other messages
+            _ => {}
+        };
+
+        Ok(message)
+    }
+
+    pub fn send(&mut self, message: &MessageType) {
+        // Create a stream to output the message
+        let mut output_stream = CodedOutputStream::new(&mut self.stream);
+
+        // Send bytes
+        output_stream.write_raw_bytes(&message.to_raw()).unwrap();
+        output_stream.flush().unwrap();
+    }
+
+    pub fn send_version(&mut self) {
+        // Send version message
+        let message = mumble::Version::new();
+        self.send(&Version(message));
+    }
+
+    pub fn send_authentication(&mut self, username: &str, password: &str) {
+        // Send authentication message
+        let mut message = mumble::Authenticate::new();
+        message.set_username(username.to_string());
+        message.set_password(password.to_string());
+        self.send(&Authenticate(message));
+    }
+
+    pub fn send_text_message(&mut self, text: &str) {
+        // Send text message
+        match self.session {
+            Some(s) => {
+                let mut message = mumble::TextMessage::new();
+                message.set_actor(s);
+                message.set_channel_id(vec![0]);
+                message.set_message(text.to_string());
+                self.send(&TextMessage(message));
+            }
+            None => println!("Connection not established. (No session)"),
+        }
+    }
+
+    pub fn change_channel(&mut self, channel_id: u32) {
+        // Send channel change message
+        if let Some(s) = self.session {
+            let mut message = mumble::UserState::new();
+            message.set_session(s);
+            message.set_actor(s);
+            message.set_channel_id(channel_id);
+            self.send(&UserState(Box::new(message)));
+        } else {
+            println!("Connection not established. (No session)");
+        }
+    }
+
+    pub fn keep_alive(&mut self) {
+        // Send ping message if necessary
+        let timestamp = SystemTime::now()
+            .duration_since(UNIX_EPOCH)
+            .unwrap()
+            .as_secs();
+
+        // Return if no new ping is necessary
+        if timestamp - self.last_ping < PING_DELAY {
+            return;
+        }
+
+        // Send ping
+        let mut ping = mumble::Ping::new();
+        ping.set_timestamp(timestamp);
+        self.last_ping = timestamp;
+        self.send(&Ping(ping));
+    }
+
+    pub fn get_users(&self) -> &Vec<mumble::UserState> {
+        &self.users
+    }
+}
+
+pub fn connect(url: &str, key: &Path, cert: &Path, username: &str, password: &str) -> Connection {
```

Comment:
The password here (and in `send_authentication`) should be an optional I think (https://mumble-protocol.readthedocs.io/en/latest/establishing_connection.html#authenticate).

Reply:
Should I just take an Option type as an argument or create two seperate constructors?